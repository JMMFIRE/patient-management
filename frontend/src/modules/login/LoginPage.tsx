import styles from './LoginPage.module.scss'
import {Button, Form} from 'react-bootstrap'
import {PatientRegisterDemoView} from '../../common/components'
import {useNavigate} from 'react-router'
import {useDispatch} from 'react-redux'
import {loginUser} from '../../features/authentication/AuthenticationSlice'
import {useEffect, useState} from 'react'
import React from 'react'
import {useLoginMutation} from '../../features/authentication/AuthenticationApiSlice'


const LoginPage = () => {
    const navigate = useNavigate()
    const dispatch = useDispatch()

    const [login, {data: loginData, error}] = useLoginMutation()
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')
    const [isInvalid, setIsInvalid] = useState(false)


    /**
     * useEffect is waiting for a change in state. When it detects that change, it'll check whether loginData exists.
     * If it does, it'll complete the login process by calling dispatch(loginUser()) and then navigate to the home page.
     *
     */
    useEffect(() => {
        if (loginData) {
            dispatch(loginUser(loginData))
            navigate('/')
        } else if (error) {
            setIsInvalid(true)
        }
    }, [error, loginData, dispatch, navigate])

    /**
     * Authenticates the user by sending the email and password to the server.
     *
     * @returns {Promise<void>}
     */
    const authenticateUser = async () => {
        const formData = {
            email,
            password
        }

        await login(formData)
    }

    return (
        <div className={styles.page}>
            <Form className={styles.loginForm} data-testid="login-form">
                <div className={styles.header}>
                    Welcome to Patient Management App
                </div>
                <div className={styles.subHeader}>
                    Login
                </div>
                <Form.Group className="mb-3" controlId="formBasicEmail">
                    <Form.Label>Email address</Form.Label>
                    <Form.Control
                        type="email"
                        placeholder="Enter email"
                        onChange={input => setEmail(input.target.value)}
                        isInvalid={isInvalid}
                        data-testid="login-email-field"
                    />
                    <Form.Control.Feedback type="invalid" data-testid="invalid-login">
                        Username or password does not match
                    </Form.Control.Feedback>
                </Form.Group>

                <Form.Group className="mb-3" controlId="formBasicPassword">
                    <Form.Label>Password</Form.Label>
                    <Form.Control
                        type="password"
                        placeholder="Password"
                        onChange={input => setPassword(input.target.value)}
                        data-testid="login-password-field"
                    />
                </Form.Group>

                <Button
                    className={styles.submitButton}
                    variant="primary"
                    type="button"
                    onClick={authenticateUser}
                    data-testid="login-button"
                >
                    Submit
                </Button>
            </Form>
            <div className={styles.demoView}>
                <div>
                    <PatientRegisterDemoView/>
                </div>
            </div>
        </div>
    )
}

export default LoginPage
