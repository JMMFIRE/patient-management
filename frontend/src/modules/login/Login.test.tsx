import React from 'react'
import '@testing-library/jest-dom'
import {renderWithProviders} from '../../common/test/test-utils'
import LoginPage from './LoginPage'
import {fireEvent, screen, waitFor} from '@testing-library/react'


describe('Login Page', () => {

    it('Renders the login page, matching snapshot', async () => {
        const loginPage = renderWithProviders(<LoginPage/>)
        expect(loginPage).toMatchSnapshot()
    })

    it('Renders the login page', async() => {
        renderWithProviders(<LoginPage/>)
        const loginForm = await screen.findByTestId('login-form')
        expect(loginForm).toBeInTheDocument()
    })

    it('Successfully logs in a user', async () => {
        const { store } = renderWithProviders(<LoginPage/>)

        const email = await screen.findByTestId('login-email-field')
        const password = await screen.findByTestId('login-password-field')
        const loginButton = await screen.findByTestId('login-button')

        fireEvent.change(email, { target: { value: 'massottoj@envisionpharma.com'}})
        fireEvent.change(password, { target: { value: 'test'}})
        fireEvent.click(loginButton)

        await waitFor(() => expect(store.getState().authentication.isAuthenticated).toBe(true))
    })

    it('Unsuccessfully logs in a user', async () => {
        renderWithProviders(<LoginPage/>)

        const email = await screen.findByTestId('login-email-field')
        const password = await screen.findByTestId('login-password-field')
        const loginButton = await screen.findByTestId('login-button')

        fireEvent.change(email, { target: { value: 'massottoj@envisionpharma.com'}})
        fireEvent.change(password, { target: { value: 'a wrong password'}})
        fireEvent.click(loginButton)

        const invalidLogin = await screen.findByTestId('invalid-login')
        expect(invalidLogin).toBeInTheDocument()
    })
})
