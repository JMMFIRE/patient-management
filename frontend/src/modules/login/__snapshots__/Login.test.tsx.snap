// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page Renders the login page, matching snapshot 1`] = `
{
  "mockedNavigate": [MockFunction],
  "renderResult": {
    "asFragment": [Function],
    "baseElement": <body>
      <div>
        <div
          class="page"
        >
          <form
            class="loginForm"
            data-testid="login-form"
          >
            <div
              class="header"
            >
              Welcome to Patient Management App
            </div>
            <div
              class="subHeader"
            >
              Login
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="formBasicEmail"
              >
                Email address
              </label>
              <input
                class="form-control"
                data-testid="login-email-field"
                id="formBasicEmail"
                placeholder="Enter email"
                type="email"
              />
              <div
                class="invalid-feedback"
                data-testid="invalid-login"
              >
                Username or password does not match
              </div>
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="formBasicPassword"
              >
                Password
              </label>
              <input
                class="form-control"
                data-testid="login-password-field"
                id="formBasicPassword"
                placeholder="Password"
                type="password"
              />
            </div>
            <button
              class="submitButton btn btn-primary"
              data-testid="login-button"
              type="button"
            >
              Submit
            </button>
          </form>
          <div
            class="demoView"
          >
            <div>
              <img
                alt="fireSpot"
                src=""
                style="width: 100%; height: 100%; object-fit: contain;"
              />
            </div>
          </div>
        </div>
      </div>
    </body>,
    "container": <div>
      <div
        class="page"
      >
        <form
          class="loginForm"
          data-testid="login-form"
        >
          <div
            class="header"
          >
            Welcome to Patient Management App
          </div>
          <div
            class="subHeader"
          >
            Login
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicEmail"
            >
              Email address
            </label>
            <input
              class="form-control"
              data-testid="login-email-field"
              id="formBasicEmail"
              placeholder="Enter email"
              type="email"
            />
            <div
              class="invalid-feedback"
              data-testid="invalid-login"
            >
              Username or password does not match
            </div>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
              for="formBasicPassword"
            >
              Password
            </label>
            <input
              class="form-control"
              data-testid="login-password-field"
              id="formBasicPassword"
              placeholder="Password"
              type="password"
            />
          </div>
          <button
            class="submitButton btn btn-primary"
            data-testid="login-button"
            type="button"
          >
            Submit
          </button>
        </form>
        <div
          class="demoView"
        >
          <div>
            <img
              alt="fireSpot"
              src=""
              style="width: 100%; height: 100%; object-fit: contain;"
            />
          </div>
        </div>
      </div>
    </div>,
    "debug": [Function],
    "findAllByAltText": [Function],
    "findAllByDisplayValue": [Function],
    "findAllByLabelText": [Function],
    "findAllByPlaceholderText": [Function],
    "findAllByRole": [Function],
    "findAllByTestId": [Function],
    "findAllByText": [Function],
    "findAllByTitle": [Function],
    "findByAltText": [Function],
    "findByDisplayValue": [Function],
    "findByLabelText": [Function],
    "findByPlaceholderText": [Function],
    "findByRole": [Function],
    "findByTestId": [Function],
    "findByText": [Function],
    "findByTitle": [Function],
    "getAllByAltText": [Function],
    "getAllByDisplayValue": [Function],
    "getAllByLabelText": [Function],
    "getAllByPlaceholderText": [Function],
    "getAllByRole": [Function],
    "getAllByTestId": [Function],
    "getAllByText": [Function],
    "getAllByTitle": [Function],
    "getByAltText": [Function],
    "getByDisplayValue": [Function],
    "getByLabelText": [Function],
    "getByPlaceholderText": [Function],
    "getByRole": [Function],
    "getByTestId": [Function],
    "getByText": [Function],
    "getByTitle": [Function],
    "queryAllByAltText": [Function],
    "queryAllByDisplayValue": [Function],
    "queryAllByLabelText": [Function],
    "queryAllByPlaceholderText": [Function],
    "queryAllByRole": [Function],
    "queryAllByTestId": [Function],
    "queryAllByText": [Function],
    "queryAllByTitle": [Function],
    "queryByAltText": [Function],
    "queryByDisplayValue": [Function],
    "queryByLabelText": [Function],
    "queryByPlaceholderText": [Function],
    "queryByRole": [Function],
    "queryByTestId": [Function],
    "queryByText": [Function],
    "queryByTitle": [Function],
    "rerender": [Function],
    "unmount": [Function],
  },
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
}
`;
